File: binding/src/libqfutureinterface.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qfutureinterfacebase.html
pub const qfutureinterfacebase = struct {

    /// New constructs a new QFutureInterfaceBase object.
    pub fn New() QtC.QFutureInterfaceBase {
        return qtc.QFutureInterfaceBase_new();
    }


    /// New2 constructs a new QFutureInterfaceBase object.
    pub fn New2(other: ?*anyopaque) QtC.QFutureInterfaceBase {
        return qtc.QFutureInterfaceBase_new2(@ptrCast(other));
    }


    /// New3 constructs a new QFutureInterfaceBase object.
    pub fn New3(initialState: i32) QtC.QFutureInterfaceBase {
        return qtc.QFutureInterfaceBase_new3(@intCast(initialState));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QFutureInterfaceBase_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn ReportStarted(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_ReportStarted(@ptrCast(self));
    }

    pub fn ReportFinished(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_ReportFinished(@ptrCast(self));
    }

    pub fn ReportCanceled(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_ReportCanceled(@ptrCast(self));
    }

    pub fn ReportException(self: ?*anyopaque, e: ?*anyopaque) void {
        qtc.QFutureInterfaceBase_ReportException(@ptrCast(self), @ptrCast(e));
    }

    pub fn ReportException2(self: ?*anyopaque, e: std::exception_ptr) void {
        qtc.QFutureInterfaceBase_ReportException2(@ptrCast(self), e);
    }

    pub fn ReportResultsReady(self: ?*anyopaque, beginIndex: i32, endIndex: i32) void {
        qtc.QFutureInterfaceBase_ReportResultsReady(@ptrCast(self), beginIndex, endIndex);
    }

    pub fn SetRunnable(self: ?*anyopaque, runnable: ?*anyopaque) void {
        qtc.QFutureInterfaceBase_SetRunnable(@ptrCast(self), @ptrCast(runnable));
    }

    pub fn SetThreadPool(self: ?*anyopaque, pool: ?*anyopaque) void {
        qtc.QFutureInterfaceBase_SetThreadPool(@ptrCast(self), @ptrCast(pool));
    }

    pub fn ThreadPool(self: ?*anyopaque, ) QtC.QThreadPool {
        return qtc.QFutureInterfaceBase_ThreadPool(@ptrCast(self));
    }

    pub fn SetFilterMode(self: ?*anyopaque, enable: bool) void {
        qtc.QFutureInterfaceBase_SetFilterMode(@ptrCast(self), enable);
    }

    pub fn SetProgressRange(self: ?*anyopaque, minimum: i32, maximum: i32) void {
        qtc.QFutureInterfaceBase_SetProgressRange(@ptrCast(self), minimum, maximum);
    }

    pub fn ProgressMinimum(self: ?*anyopaque, ) i32 {
        return qtc.QFutureInterfaceBase_ProgressMinimum(@ptrCast(self));
    }

    pub fn ProgressMaximum(self: ?*anyopaque, ) i32 {
        return qtc.QFutureInterfaceBase_ProgressMaximum(@ptrCast(self));
    }

    pub fn IsProgressUpdateNeeded(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsProgressUpdateNeeded(@ptrCast(self));
    }

    pub fn SetProgressValue(self: ?*anyopaque, progressValue: i32) void {
        qtc.QFutureInterfaceBase_SetProgressValue(@ptrCast(self), progressValue);
    }

    pub fn ProgressValue(self: ?*anyopaque, ) i32 {
        return qtc.QFutureInterfaceBase_ProgressValue(@ptrCast(self));
    }

    pub fn SetProgressValueAndText(self: ?*anyopaque, progressValue: i32, progressText: []const u8) void {
        const progressText_str = qtc.libqt_string{
    .len = progressText.len,
    .data = progressText.ptr,
};
qtc.QFutureInterfaceBase_SetProgressValueAndText(@ptrCast(self), progressValue, progressText_str);
    }

    pub fn ProgressText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFutureInterfaceBase_ProgressText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfutureinterfacebase.ProgressText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetExpectedResultCount(self: ?*anyopaque, resultCount: i32) void {
        qtc.QFutureInterfaceBase_SetExpectedResultCount(@ptrCast(self), resultCount);
    }

    pub fn ExpectedResultCount(self: ?*anyopaque, ) i32 {
        return qtc.QFutureInterfaceBase_ExpectedResultCount(@ptrCast(self));
    }

    pub fn ResultCount(self: ?*anyopaque, ) i32 {
        return qtc.QFutureInterfaceBase_ResultCount(@ptrCast(self));
    }

    pub fn QueryState(self: ?*anyopaque, state: i32) bool {
        return qtc.QFutureInterfaceBase_QueryState(@ptrCast(self), @intCast(state));
    }

    pub fn IsRunning(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsRunning(@ptrCast(self));
    }

    pub fn IsStarted(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsStarted(@ptrCast(self));
    }

    pub fn IsCanceled(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsCanceled(@ptrCast(self));
    }

    pub fn IsFinished(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsFinished(@ptrCast(self));
    }

    pub fn IsPaused(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsPaused(@ptrCast(self));
    }

    pub fn SetPaused(self: ?*anyopaque, paused: bool) void {
        qtc.QFutureInterfaceBase_SetPaused(@ptrCast(self), paused);
    }

    pub fn TogglePaused(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_TogglePaused(@ptrCast(self));
    }

    pub fn IsSuspending(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsSuspending(@ptrCast(self));
    }

    pub fn IsSuspended(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsSuspended(@ptrCast(self));
    }

    pub fn IsThrottled(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsThrottled(@ptrCast(self));
    }

    pub fn IsResultReadyAt(self: ?*anyopaque, index: i32) bool {
        return qtc.QFutureInterfaceBase_IsResultReadyAt(@ptrCast(self), index);
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsValid(@ptrCast(self));
    }

    pub fn LoadState(self: ?*anyopaque, ) i32 {
        return qtc.QFutureInterfaceBase_LoadState(@ptrCast(self));
    }

    pub fn Cancel(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_Cancel(@ptrCast(self));
    }

    pub fn CancelAndFinish(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_CancelAndFinish(@ptrCast(self));
    }

    pub fn SetSuspended(self: ?*anyopaque, suspendVal: bool) void {
        qtc.QFutureInterfaceBase_SetSuspended(@ptrCast(self), suspendVal);
    }

    pub fn ToggleSuspended(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_ToggleSuspended(@ptrCast(self));
    }

    pub fn ReportSuspended(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_ReportSuspended(@ptrCast(self));
    }

    pub fn SetThrottled(self: ?*anyopaque, enable: bool) void {
        qtc.QFutureInterfaceBase_SetThrottled(@ptrCast(self), enable);
    }

    pub fn WaitForFinished(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_WaitForFinished(@ptrCast(self));
    }

    pub fn WaitForNextResult(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_WaitForNextResult(@ptrCast(self));
    }

    pub fn WaitForResult(self: ?*anyopaque, resultIndex: i32) void {
        qtc.QFutureInterfaceBase_WaitForResult(@ptrCast(self), resultIndex);
    }

    pub fn WaitForResume(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_WaitForResume(@ptrCast(self));
    }

    pub fn SuspendIfRequested(self: ?*anyopaque, ) void {
        qtc.QFutureInterfaceBase_SuspendIfRequested(@ptrCast(self));
    }

    pub fn Mutex(self: ?*anyopaque, ) QtC.QMutex {
        return qtc.QFutureInterfaceBase_Mutex(@ptrCast(self));
    }

    pub fn HasException(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_HasException(@ptrCast(self));
    }

    pub fn ExceptionStore(self: ?*anyopaque, ) QtC.QtPrivate__ExceptionStore {
        return qtc.QFutureInterfaceBase_ExceptionStore(@ptrCast(self));
    }

    pub fn ResultStoreBase(self: ?*anyopaque, ) QtC.QtPrivate__ResultStoreBase {
        return qtc.QFutureInterfaceBase_ResultStoreBase(@ptrCast(self));
    }

    pub fn ResultStoreBase2(self: ?*anyopaque, ) QtC.QtPrivate__ResultStoreBase {
        return qtc.QFutureInterfaceBase_ResultStoreBase2(@ptrCast(self));
    }

    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QFutureInterfaceBase_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QFutureInterfaceBase_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QFutureInterfaceBase_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsChainCanceled(self: ?*anyopaque, ) bool {
        return qtc.QFutureInterfaceBase_IsChainCanceled(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QFutureInterfaceBase_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/futureinterface.html#types
pub const enums = struct {
    pub const State = enum {
        pub const NoState: i32 = 0;
        pub const Running: i32 = 1;
        pub const Started: i32 = 2;
        pub const Finished: i32 = 4;
        pub const Canceled: i32 = 8;
        pub const Suspending: i32 = 16;
        pub const Suspended: i32 = 32;
        pub const Throttled: i32 = 64;
        pub const Pending: i32 = 128;
    };

    pub const CancelMode = enum {
        pub const CancelOnly: i32 = 0;
        pub const CancelAndFinish: i32 = 1;
    };

};

File: binding/src/libqfutureinterface.cpp
#include <QFutureInterfaceBase>
#include <QMutex>
#include <QRunnable>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QThreadPool>
#include <qfutureinterface.h>
#include "libqfutureinterface.h"
#include "libqfutureinterface.hxx"

QFutureInterfaceBase* QFutureInterfaceBase_new() {
	 return new QFutureInterfaceBase();
}

QFutureInterfaceBase* QFutureInterfaceBase_new2(const QFutureInterfaceBase* other) {
	 return new QFutureInterfaceBase(*other);
}

QFutureInterfaceBase* QFutureInterfaceBase_new3(int initialState) {
	 return new QFutureInterfaceBase(static_cast<QFlags<QStyle::StateFlag>>(initialState));
}

void QFutureInterfaceBase_OperatorAssign(QFutureInterfaceBase* self, const QFutureInterfaceBase* other) {
	self->operator=(*other);
}

void QFutureInterfaceBase_ReportStarted(QFutureInterfaceBase* self) {
	self->reportStarted();
}

void QFutureInterfaceBase_ReportFinished(QFutureInterfaceBase* self) {
	self->reportFinished();
}

void QFutureInterfaceBase_ReportCanceled(QFutureInterfaceBase* self) {
	self->reportCanceled();
}

void QFutureInterfaceBase_ReportResultsReady(QFutureInterfaceBase* self, int beginIndex, int endIndex) {
	self->reportResultsReady(beginIndex, endIndex);
}

void QFutureInterfaceBase_SetRunnable(QFutureInterfaceBase* self, QRunnable* runnable) {
	self->setRunnable(runnable);
}

void QFutureInterfaceBase_SetThreadPool(QFutureInterfaceBase* self, QThreadPool* pool) {
	self->setThreadPool(pool);
}

QThreadPool* QFutureInterfaceBase_ThreadPool(const QFutureInterfaceBase* self) {
	return self->threadPool();
}

void QFutureInterfaceBase_SetFilterMode(QFutureInterfaceBase* self, bool enable) {
	self->setFilterMode(enable);
}

void QFutureInterfaceBase_SetProgressRange(QFutureInterfaceBase* self, int minimum, int maximum) {
	self->setProgressRange(minimum, maximum);
}

int QFutureInterfaceBase_ProgressMinimum(const QFutureInterfaceBase* self) {
	return self->progressMinimum();
}

int QFutureInterfaceBase_ProgressMaximum(const QFutureInterfaceBase* self) {
	return self->progressMaximum();
}

bool QFutureInterfaceBase_IsProgressUpdateNeeded(const QFutureInterfaceBase* self) {
	return self->isProgressUpdateNeeded();
}

void QFutureInterfaceBase_SetProgressValue(QFutureInterfaceBase* self, int progressValue) {
	self->setProgressValue(progressValue);
}

int QFutureInterfaceBase_ProgressValue(const QFutureInterfaceBase* self) {
	return self->progressValue();
}

void QFutureInterfaceBase_SetProgressValueAndText(QFutureInterfaceBase* self, int progressValue, const libqt_string progressText) {
	self->setProgressValueAndText(progressValue, QString::fromUtf8(progressText.data, progressText.len));
}

libqt_string QFutureInterfaceBase_ProgressText(const QFutureInterfaceBase* self) {
	QString _ret = self->progressText();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QFutureInterfaceBase_SetExpectedResultCount(QFutureInterfaceBase* self, int resultCount) {
	self->setExpectedResultCount(resultCount);
}

int QFutureInterfaceBase_ExpectedResultCount(QFutureInterfaceBase* self) {
	return self->expectedResultCount();
}

int QFutureInterfaceBase_ResultCount(const QFutureInterfaceBase* self) {
	return self->resultCount();
}

bool QFutureInterfaceBase_QueryState(const QFutureInterfaceBase* self, int state) {
	return self->queryState(static_cast<QFlags<QStyle::StateFlag>>(state));
}

bool QFutureInterfaceBase_IsRunning(const QFutureInterfaceBase* self) {
	return self->isRunning();
}

bool QFutureInterfaceBase_IsStarted(const QFutureInterfaceBase* self) {
	return self->isStarted();
}

bool QFutureInterfaceBase_IsCanceled(const QFutureInterfaceBase* self) {
	return self->isCanceled();
}

bool QFutureInterfaceBase_IsFinished(const QFutureInterfaceBase* self) {
	return self->isFinished();
}

bool QFutureInterfaceBase_IsPaused(const QFutureInterfaceBase* self) {
	return self->isPaused();
}

void QFutureInterfaceBase_SetPaused(QFutureInterfaceBase* self, bool paused) {
	self->setPaused(paused);
}

void QFutureInterfaceBase_TogglePaused(QFutureInterfaceBase* self) {
	self->togglePaused();
}

bool QFutureInterfaceBase_IsSuspending(const QFutureInterfaceBase* self) {
	return self->isSuspending();
}

bool QFutureInterfaceBase_IsSuspended(const QFutureInterfaceBase* self) {
	return self->isSuspended();
}

bool QFutureInterfaceBase_IsThrottled(const QFutureInterfaceBase* self) {
	return self->isThrottled();
}

bool QFutureInterfaceBase_IsResultReadyAt(const QFutureInterfaceBase* self, int index) {
	return self->isResultReadyAt(index);
}

bool QFutureInterfaceBase_IsValid(const QFutureInterfaceBase* self) {
	return self->isValid();
}

int QFutureInterfaceBase_LoadState(const QFutureInterfaceBase* self) {
	return self->loadState();
}

void QFutureInterfaceBase_Cancel(QFutureInterfaceBase* self) {
	self->cancel();
}

void QFutureInterfaceBase_CancelAndFinish(QFutureInterfaceBase* self) {
	self->cancelAndFinish();
}

void QFutureInterfaceBase_SetSuspended(QFutureInterfaceBase* self, bool suspendVal) {
	self->setSuspended(suspendVal);
}

void QFutureInterfaceBase_ToggleSuspended(QFutureInterfaceBase* self) {
	self->toggleSuspended();
}

void QFutureInterfaceBase_ReportSuspended(const QFutureInterfaceBase* self) {
	self->reportSuspended();
}

void QFutureInterfaceBase_SetThrottled(QFutureInterfaceBase* self, bool enable) {
	self->setThrottled(enable);
}

void QFutureInterfaceBase_WaitForFinished(QFutureInterfaceBase* self) {
	self->waitForFinished();
}

bool QFutureInterfaceBase_WaitForNextResult(QFutureInterfaceBase* self) {
	return self->waitForNextResult();
}

void QFutureInterfaceBase_WaitForResult(QFutureInterfaceBase* self, int resultIndex) {
	self->waitForResult(resultIndex);
}

void QFutureInterfaceBase_WaitForResume(QFutureInterfaceBase* self) {
	self->waitForResume();
}

void QFutureInterfaceBase_SuspendIfRequested(QFutureInterfaceBase* self) {
	self->suspendIfRequested();
}

QMutex* QFutureInterfaceBase_Mutex(const QFutureInterfaceBase* self) {
	return new QMutex(self->mutex());
}

bool QFutureInterfaceBase_HasException(const QFutureInterfaceBase* self) {
	return self->hasException();
}

bool QFutureInterfaceBase_OperatorEqual(const QFutureInterfaceBase* self, const QFutureInterfaceBase* other) {
	return self->operator==(*other);
}

bool QFutureInterfaceBase_OperatorNotEqual(const QFutureInterfaceBase* self, const QFutureInterfaceBase* other) {
	return self->operator!=(*other);
}

void QFutureInterfaceBase_Swap(QFutureInterfaceBase* self, QFutureInterfaceBase* other) {
	self->swap(*other);
}

bool QFutureInterfaceBase_IsChainCanceled(const QFutureInterfaceBase* self) {
	return self->isChainCanceled();
}

void QFutureInterfaceBase_Delete(QFutureInterfaceBase* self) {
    delete self;
}

